pymc/distributions/mixture.py

```diff
@@ -53,516 +72,357 @@ class Mixture(Distribution):
+
+            # The mixture is an array of 5 elements
+            # Each element can be thought of as an independent scalar mixture of 3
+            # components with different means
+            like = pm.Mixture('like', w=w, comp_dists=components, observed=data)
+
 
-            w = pm.Dirichlet('w', a=np.ones(3, 5))  # w is a stack of 3 independent 5 component weight arrays
+    .. code-block:: python
+
+        # Mixture of 2 Dirichlet variables
+        with pm.Model() as model:
+            w = pm.Dirichlet('w', a=np.ones(2))  # 2 mixture weights
 
-            components = pm.Normal.dist(mu=mu, sigma=1, shape=(3, 5))
+            # These two forms are equivalent, but the second benefits from vectorization
+            components = [
+                pm.Dirichlet.dist(a=[1, 10, 100], shape=(3,)),
+                pm.Dirichlet.dist(a=[100, 10, 1], shape=(3,)),
+            ]
+            components = pm.Dirichlet.dist(a=[[1, 10, 100], [100, 10, 1]], shape=(2, 3))
 
-            # The mixture is an array of 3 elements.
-            # Each can be thought of as an independent scalar mixture of 5 components
-            like = pm.Mixture('like', w=w, comp_dists = components, observed=data, shape=3)
+            # The mixture is an array of 3 elements
+            # Each element comes from only one of the two core Dirichlet components
+            like = pm.Mixture('like', w=w, comp_dists=components, observed=data)
     """
 
-    def __init__(self, w, comp_dists, *args, **kwargs):
-        # comp_dists type checking
-        if not (
-            isinstance(comp_dists, Distribution)
-            or (
-                isinstance(comp_dists, Iterable)
-                and all(isinstance(c, Distribution) for c in comp_dists)
+    @classmethod
+    def dist(cls, w, comp_dists, **kwargs):
+        if not isinstance(comp_dists, (tuple, list)):
+            # comp_dists is a single component
+            comp_dists = [comp_dists]
+        elif len(comp_dists) == 1:
+            warnings.warn(
+                "Single component will be treated as a mixture across the last size dimension.\n"
+                "To disable this warning do not wrap the single component inside a list or tuple",
+                UserWarning,
             )
-        ):
-            raise TypeError(
-                "Supplied Mixture comp_dists must be a "
-                "Distribution or an iterable of "
-                "Distributions. Got {} instead.".format(
-                    type(comp_dists)
-                    if not isinstance(comp_dists, Iterable)
-                    else [type(c) for c in comp_dists]
+
+        # Check that components are not associated with a registered variable in the model
+        components_ndim = set()
+        components_ndim_supp = set()
+        for dist in comp_dists:
+            # TODO: Allow these to not be a RandomVariable as long as we can call `ndim_supp` on them
+            #  and resize them
```

Comment:
Just wanting to clarify (again?) for my understanding: what else can `dist` be? For instance, is this where we want to be more flexible in allowing things like nested mixtures and censored distributions? E.g. something like [line 208 below](https://github.com/pymc-devs/pymc/pull/5438/files#diff-a46f0cbffb99eea1c2c7c51bfae351225ae8f38ad9b825d0b4687a7456bc6288R208)

Reply:
Yeah exactly!